staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -476,6 +481,20 @@ func (r Request) finalURLTemplate() url.URL {
 	return *url
 }
 
+// replaceNamespace4DynamicClient replace the actual client namespace string
+// with "{namespace}" for dynamic client, as the namespace string is appended
+// in r.pathPrefix
+func replaceNamespace4DynamicClient(oldPath string) string {
+	segments := strings.Split(oldPath, "/")
+	for i := 0; i < len(segments)-1; i++ {
+		if segments[i] == "namespaces" {
```

Comment:
One concern is:
 - if the resource [name](https://github.com/kubernetes/kubernetes/blob/45c5ddd2fb8f10e097483257663fa6a7be59ffa9/staging/src/k8s.io/client-go/dynamic/simple.go#L194) is "namespaces" (e.g. a node called "namespaces");
 - if [r.baseURL.Path](https://github.com/kubernetes/kubernetes/blob/45c5ddd2fb8f10e097483257663fa6a7be59ffa9/staging/src/k8s.io/client-go/rest/request.go#L275) contains "/namespaces/"

I'm not sure if we have constraints (e.g. resources cannot be named "namespaces") that could help us rule out the two cases above. If not, you could check if [r.pathPrefix](https://github.com/kubernetes/kubernetes/blob/45c5ddd2fb8f10e097483257663fa6a7be59ffa9/staging/src/k8s.io/client-go/rest/request.go#L90) contains [r.baseURL.Path](https://github.com/kubernetes/kubernetes/blob/45c5ddd2fb8f10e097483257663fa6a7be59ffa9/staging/src/k8s.io/client-go/rest/request.go#L85), and ignore the baseURL part during parsing. The rest of pathPrefix should follow Kubernetes API convention format (legacy group v.s. named group; cluster-scoped v.s. namespaced). 

This also helps you parse the {name} segment.

Reply:
Another case is the namespace kind, where the url is `/api/v1/namespaces/{name}`

Reply:
I checked in my local cluster, it seems like I can create a namespace called "namespace" with no problem. But in that case, the replaceNamespace4DynamicClient function can still handle it by blindly replace whatever segment is following /namespaces/ with {namespace}.  [r.baseURL](https://github.com/kubernetes/kubernetes/blob/426ef9d349bb3a277c3e8826fb772b6bdb008382/staging/src/k8s.io/client-go/rest/request.go#L85) and [pathPrefix](https://github.com/kubernetes/kubernetes/blob/426ef9d349bb3a277c3e8826fb772b6bdb008382/staging/src/k8s.io/client-go/rest/request.go#L90) are separate in rest.Request. So I don't get how would checking r.pathPrefix and r.baseURL.Path help? 

For example, as in testcase now, if full url is "http://localhost/pre1/namespaces/ns/r1", then r.pathPrefix is "/pre1/namespaces/ns/r1". r.baseURL is "http://localhost"

Reply:
I am a little skeptical of this approach, but maybe you can make it work. :)

You at least need to check more than one path segment. Here's some test cases off the top of my head.

* `/api/v1/namespaces/kube-system/services/monitoring-heapster` is about a service; it should match.
* `/api/v1/namespaces/kube-system` is about a namespace; it should NOT match (kube-system is the name of the resource in this case, which is not inside a namespace).
* `/apis/apps/v1/namespaces/foo/deployments/bar` is about resource `bar` in namespace `foo`, it should match.
* `/apis/apps/v1/namespaces/namespaces/deployments/namespaces` is about resource `namespaces` in namespace `namespaces`, it should match.

Reply:
I just pushed a new commit after talking to @roycaihw offline. It parses the request.pathPrefix based on the type of API groups, either in /api/v1... format for legacy group or /apis/apps/v1... format for named group. 

Based on @lavalamp comment, I think there is no rule on nomenclature rule... As far as Daniel's concern, the first, third and forth case should all the fine. However, both Haowei and Daniel mentioned that if we have a dynamic resource client that has no namespace but has a resource called "namespaces", then current fix would blindly replace the segment after the resource name with "{namespace}", which is wrong...

In that case, we will need to pass the [dynamicResourceClient](https://github.com/kubernetes/kubernetes/blob/45c5ddd2fb8f10e097483257663fa6a7be59ffa9/staging/src/k8s.io/client-go/dynamic/simple.go#L69) struct to rest/request.go so it knows exactly what the structure is instead of blind string manipulation. And it would be more than "only have to do the analysis if you have no hits for the namespace being set." as @deads2k [original suggested](https://github.com/kubernetes/kubernetes/issues/68115#issuecomment-418496136), let me do more analysis and see if it would have impact on other parts than just metrics.

Reply:
> `/api/v1/namespaces/kube-system` is about a namespace; it should NOT match

Given the fact that user can choose to [set pathPrefix, resource and resourceName at the same time](https://github.com/kubernetes/kubernetes/blob/ae8a046985d1e768dc589509d584b6fef13e900a/staging/src/k8s.io/client-go/rest/request.go#L421-L433), it can also be part of a `/api/v1/namespaces/kube-system/services/monitoring-heapster` request, in which case it should match. 

We either need to involve more fields in [Request](https://github.com/kubernetes/kubernetes/blob/ae8a046985d1e768dc589509d584b6fef13e900a/staging/src/k8s.io/client-go/rest/request.go#L80-L101) into the detection logic, or use some other approach (e.g. modify the dynamic client)